### PR TITLE
remove duplicate accounts in chain-spec.

### PR DESCRIPTION
**Pull request by kianenigma**
as reported by https://github.com/paritytech/substrate/issues/7713, almost insubstantial.

### CODE REVIEW

**Review by bkchr — CHANGES_REQUESTED** [review]
Not sure what this change should bring as value?

Users of Substrate will get no benefit of this. The balances pallet should check this and throw an error if there are duplicated accounts in genesis.

**Comment by kianenigma** [comment]
I think it is reasonable for one to argue that balances does not care about this and in case of duplicate just overrides the previous value, but I can do as you recommend.

**Comment by thiolliere** [comment]
I also prefer having balance to care about overrides than having some check node/cli

**Comment by bkchr** [comment]
I don't see any reason why someone should add an account twice, when not done by accident.

Ty for fixing this.

**Review by shawntabrizi — APPROVED** [review]

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -412,6 +412,14 @@ decl_storage! {
 					"the balance of any account should always be at least the existential deposit.",
 				)
 			}
+
+			// ensure no duplicates exist.
+			let mut endowed_accounts = config.balances.iter().map(|(x, _)| x).cloned().collect::<Vec<_>>();
```
Comment (bkchr) [inline frame/balances/src/lib.rs:417]:
```suggestion
			let endowed_accounts = config.balances.iter().map(|(x, _)| x).cloned().collect::<std::collections::HashSet<_>>();
```

File: frame/balances/src/lib.rs [hunk 2]
```diff
@@ -412,6 +412,14 @@ decl_storage! {
...
+			endowed_accounts.sort();
```
Comment (bkchr) [inline frame/balances/src/lib.rs:418]:
```suggestion
```

File: frame/balances/src/lib.rs [hunk 3]
```diff
@@ -412,6 +412,14 @@ decl_storage! {
...
+			endowed_accounts.dedup();
```
Comment (bkchr) [inline frame/balances/src/lib.rs:419]:
```suggestion
```

**Review by bkchr — APPROVED** [review]

File: frame/balances/src/lib.rs (hunk 1 above)
Comment (kianenigma) [reply frame/balances/src/lib.rs]: Ahh this is very nice, but I think it borked the build Sir :P

File: frame/balances/src/lib.rs [hunk 4]
```diff
@@ -412,6 +412,12 @@ decl_storage! {
 					"the balance of any account should always be at least the existential deposit.",
 				)
 			}
+
+			// ensure no duplicates exist.
+			let endowed_accounts = config.balances.iter().map(|(x, _)| x).cloned().collect::<std::collections::HashSet<_>>();
```
Comment (bkchr) [inline frame/balances/src/lib.rs:417]:
```suggestion
			let endowed_accounts = config.balances.iter().map(|(x, _)| x).cloned().collect::<std::collections::BTreeSet<_>>();
```

**Comment by bkchr** [comment]
bot merge

**Comment by shawntabrizi** [comment]
You can force merge this to avoid having to rebuild everything

**Comment by kianenigma** [comment]
Not sure what the CI complain is about tbh.

**Comment by bkchr** [comment]
bot force merge

**Comment by bkchr** [comment]
bot merge force

**Comment by bkchr** [comment]
bot merge

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**